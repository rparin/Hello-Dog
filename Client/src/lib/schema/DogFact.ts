import { z } from "zod";

export const DogFactSchema = z.strictObject({
  dogFact: z.string(),
});
