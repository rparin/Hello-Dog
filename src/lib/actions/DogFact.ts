"use server";

export async function getDogFact() {
  const res = await fetch("https://dogapi.dog/api/v2/facts", {
    method: "GET",
    headers: {
      "Content-Type": "application/vnd.api+json; charset=utf-8",
    },
    cache: "no-store",
  });
  return await res.json();
}
