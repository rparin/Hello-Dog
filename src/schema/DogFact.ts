import { z } from "zod";

export const DogFactSchema = z.object({
  data: z.array(
    z.strictObject({
      id: z.string(),
      type: z.literal("fact"),
      attributes: z.object({
        body: z.string(),
      }),
    })
  ),
});
