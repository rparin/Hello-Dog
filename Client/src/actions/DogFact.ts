"use server";
import { parsedEnv } from "@env/env";
import { DogFactSchema } from "@/schema/DogFact";
import { fromError } from "zod-validation-error";

export async function getDogFact(len: number = 120) {
  var res;

  try {
    res = await fetch(`${parsedEnv.SERVER_URL}/api/dogfact?max_length=${len}`, {
      method: "GET",
      headers: {
        "Content-Type": "application/json",
      },
      cache: "no-store",
    });
  } catch (error) {
    throw new Error("Dog Fact: Fetch Error", { cause: error });
  }

  if (!res?.ok) {
    throw new Error("Dog Fact: Response Error", {
      cause: `HTTP Response Code: ${res?.status}`,
    });
  }

  const data = await res.json();
  const result = DogFactSchema.safeParse(data);
  if (!result.success) {
    throw new Error("Dog Fact: Invalid API Response format", {
      cause: fromError(result.error).toString(),
    });
  } else {
    return result.data.dogFact;
  }
}
