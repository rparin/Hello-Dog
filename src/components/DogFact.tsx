import { cn } from "@/lib/utils";
import { ExternalLink } from "lucide-react";
import { getDogFact } from "@/lib/actions/DogFact";
import { delay } from "@/lib/utils";

export default async function DogFact({ className }: { className?: string }) {
  var dogFact =
    'Upright, stiff, rapid tail movement is not wagging or "friendly" but indicates a dog who\'s rather excited and focused.';

  for (let i = 0; i < 3; i++) {
    var tempDogFact = await getDogFact();
    if (tempDogFact.length < dogFact.length) {
      dogFact = tempDogFact;
      break;
    }
    await delay(200);
  }

  return (
    <div
      className={cn(
        "relative flex h-full flex-col justify-center text-start text-black md:gap-0.5 lg:gap-1",
        className
      )}>
      <div className="absolute right-3 top-0.5 z-50 flex w-full justify-end gap-1 text-xs opacity-60 lg:right-5">
        <p className="leading-none">Powered by Stratonauts Dog API</p>
        <a
          href="https://dogapi.dog"
          aria-label="Stratonauts Dog API"
          target="_blank"
          rel="noopener noreferrer">
          <ExternalLink className="h-[.6rem] w-auto" />
        </a>
      </div>
      <h1 className="mx-2 text-start text-base leading-none underline md:mx-3 lg:text-center lg:text-xl xl:mb-0.5">
        Random Dog Fact
      </h1>
      <p className="mx-2 line-clamp-1 text-sm leading-none md:mx-3 lg:mx-4 lg:text-base">
        {dogFact}
      </p>
    </div>
  );
}
