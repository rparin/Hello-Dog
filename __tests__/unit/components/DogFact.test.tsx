import { render } from "@testing-library/react";
import { axe, toHaveNoViolations } from "jest-axe";
import DogFact from "@/components/DogFact";
import mockFetch from "@test-utils/fetchMock";

expect.extend(toHaveNoViolations);

test("it should have no accessibility violations", async () => {
  //Mock get request
  //https://dogapi.dog/api/v2/facts
  window.fetch = mockFetch({
    data: [
      {
        id: "115fabfe-ea24-42af-afde-2cdcad7a8502",
        type: "fact",
        attributes: {
          body: "Tired puppies get cranky just like little kids. If you have a fussy puppy, try nap time.",
        },
      },
    ],
  });

  const { container } = render(await DogFact({ className: "" }));
  const results = await axe(container);
  expect(results).toHaveNoViolations();
});
