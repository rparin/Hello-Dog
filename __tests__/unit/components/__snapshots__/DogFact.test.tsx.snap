// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it renders component unchanged 1`] = `
<div>
  <div
    class="relative flex h-full flex-col justify-center text-start text-black md:gap-0.5 lg:gap-1"
  >
    <div
      class="absolute right-3 top-0.5 z-50 flex w-full justify-end gap-1 text-xs opacity-60 lg:right-5"
    >
      <p
        class="leading-none"
      >
        Powered by Stratonauts Dog API
      </p>
      <a
        aria-label="Stratonauts Dog API"
        href="https://dogapi.dog"
        rel="noopener noreferrer"
        target="_blank"
      >
        <svg
          class="lucide lucide-external-link h-[.6rem] w-auto"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M15 3h6v6"
          />
          <path
            d="M10 14 21 3"
          />
          <path
            d="M18 13v6a2 2 0 0 1-2 2H5a2 2 0 0 1-2-2V8a2 2 0 0 1 2-2h6"
          />
        </svg>
      </a>
    </div>
    <h1
      class="mx-2 text-start text-base leading-none underline md:mx-3 lg:text-center lg:text-xl xl:mb-0.5"
    >
      Random Dog Fact
    </h1>
    <p
      class="mx-2 text-sm leading-none md:mx-3 lg:mx-4 lg:text-base"
    >
      Tired puppies get cranky just like little kids. If you have a fussy puppy, try nap time.
    </p>
  </div>
</div>
`;
